src/core/execute.h

```diff
@@ -221,6 +221,15 @@ struct ExecContext {
         mode_t runtime_directory_mode;
         ExecPreserveMode runtime_directory_preserve_mode;
 
+        char **data_directory;
+        mode_t data_directory_mode;
+
+        char **cache_directory;
+        mode_t cache_directory_mode;
+
+        char **log_directory;
+        mode_t log_directory_mode;
```

Comment:
so, give that we have four times very similar fields here I think we should introduce a new enum:

```c
enum ExecDirectoryType {
        EXEC_DIRECTORY_RUNTIME = 0,
        EXEC_DIRECTORY_STATE,
        EXEC_DIRECTORY_CACHE,
        EXEC_DIRECTORY_LOG,
        _EXEC_DIRECTORY_MAX,
        _EXEC_DIRECTORY_INVALID = -1,
};
```

And then we can define a structure for the combination of log directories and the mode:

```c
typedef struct ExecDirectory {
        char **paths;
        mode_t mode;
};
```

and finally, we can then just have a single array for all these directories as part of ExecContext: 

```c
        …
        ExecDirectory directories[_EXEC_DIRECTORY_MAX];
        …
```

Given that in most cases we need to treat the four directories in very similar ways we can then easily iterate over all four types. And if we need one dir specifically we can used `exec_context->directories[EXEC_CONTEXT_RUNTIME]` and so on specifically.

Does that make sense?